internal/service/dataset/iterator.go

```diff
@@ -0,0 +1,165 @@
+// Copyright 2022 MIMIRO AS
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dataset
+
+import (
+	"encoding/binary"
+	"fmt"
+	"io"
+
+	"github.com/dgraph-io/badger/v3"
+
+	"github.com/mimiro-io/datahub/internal/service/store"
+)
+
+type IterableDataset interface {
+	At(since uint64) (Iterator, error)
+}
+
+type IterableBadgerDataset struct {
+	db        *badger.DB
+	datasetID uint32
+}
+type Iterator interface {
+	io.Closer
+	Inverse() Iterator
+	LatestOnly() Iterator
+	Next() bool
+	Item() []byte
+	NextOffset() uint64
+	Error() error
+}
+type BadgerDatasetIterator struct {
+	db             *badger.DB
+	err            error
+	txn            *badger.Txn
+	it             *badger.Iterator
+	datasetID      uint32
+	startingOffset uint64
+	datasetPrefix  []byte
+	item           []byte
+	offset         uint64
+	inverse        bool
+	latestOnly     bool
+}
+
+func (b *BadgerDatasetIterator) Close() error {
+	if b.txn != nil {
+		defer b.txn.Discard()
+	}
+	if b.it != nil {
+		b.it.Close()
+	}
+	return nil
+}
+
+func (b *BadgerDatasetIterator) Inverse() Iterator {
```

Comment:
Oh, I understand it, you use the call to Inverse() to switch the order. Still not sure that's the best way.